[14.0] Release

ivs-cetmix opened the pull request:
And here we go!
 🚀 🚀 🚀 🚀 🚀 

Merge dev into main.

Task: 4513

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Advanced remote server management and command execution capabilities with a streamlined, revitalized UI.
  - Enhanced Git project management with multi-language support and integrated flight plan and file synchronization functionalities.
  - New access control rules for managing permissions across various models including commands, files, and keys.
  - Introduction of new models for managing server templates, SSH keys, command logs, and flight plans.
  - New templates for bug reports and feature requests to streamline community feedback.
  - Enhanced configuration options related to GitHub CI and Odoo settings.
  - New documentation files for the Cetmix Tower and Cetmix Tower Server, detailing functionality and usage.

- **Documentation**
  - Comprehensive updates to user-facing guides including README, USAGE, CONTRIBUTING, and CONFIGURE materials.
  - New issue and feature request templates to streamline community feedback.

- **Chores**
  - Improvements to CI/CD workflows and configuration settings for increased code quality and deployment stability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Aldeigja committed:
[IMP] cetmix_tower_server: managers can delete own plan line actions

        Before this commit
        ------------------
        - Members of the group_manager cannot delete the plan line actions they have created.
        - Members of the manager and user groups can access plan line actions assigned
	 to plans belonging to servers to which these users are not subscribed.

        After this commit
        ------------------
       - Members of the group_manager can delete the plan line actions they have created if
         plans which these assigned doesn't belong to any server or belong to server
         to which this user subscribed
       - Members of the manager and user groups can not access plan line actions assigned
         to plans belonging to servers to which these users are not subscribed

Commit:
[IMP] cetmix_tower_server: managers can delete own plan lines
Before this commit
------------------

- Members of the group_manager cannot delete the plan lines they have created
- Members of the manager and user groups can access plan lines assigned
  to plans belonging to servers to which these users are not subscribed.

After this commit
------------------
- Members of the group_manager can delete the plan lines they have created if
  plans which these assigned doesn't belong to any server or belong to server
  to which this user subscribed
- Members of the manager and user groups can not access plan lines assigned
  to plans belonging to servers to which these users are not subscribed.

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules:
Before this commit:
memnbers of group_user and group_manager have no access to variable values related to plan line actions

After this commit:
members of group user can read variable values related to plan line actions
members of group manager can
-  read variable values related to plan line actions
-  delete variable values related to plan line actions  they have created

Commit:
[FIX] cetmix_tower_server Access mixin

cx.tower.access.mixin

Before this commit
------------------

'access_level' field was restricted to 'Manager' and 'Root' groups only.
It caused access error when member 'User' group tried to access it.
Although it appeared only in tests it could have created potential
issues in other cases.

After this commit
-----------------

Remove group restriction from 'access_level' field.
If needed they can be applied at the view level.

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules in pline line actions:
Before this commit
------------------

    group_user and group_manager users have access to the variable values  associated
    with the plan line actions regardless of whether or not they were or were not
    subscribed to the servers associated with the parent plans for those plan line actions

After this commi
---------------

If they are subscribed to the servers associated to plans parent for these actions, then:

    members of group user can read variable values related to plan line actions
    members of group manager can
    -  read snd update variable values related to plan line actions
    -  delete variable values they have created

Aldeigja committed:
[IMP] cetmix_tower_server: update variable values access rules in templates:
    Before this commit
    ------------------

        members of group_manager don't have access to the variable values associated
        with the server templates.

    After this commit
    ---------------

    If members of group manager  are subscribed to the server template then:
    members of group manager can
        -  read and update variable values related to server templates they are subscribed
        -  delete variable values they have created

tendil committed:
[IMP] cetmix_tower_server: enhance and validate required vars

This commit consolidates and enhances handling of required
configuration variables in server templates and the "Create Server" wizard.
It includes validation improvements, backend logic adjustments,
test reliability fixes, and translation updates.

Key changes:
- Introduced validation for required variables during server creation:
  - Servers cannot be created without valid values for required variables.
  - Optional variables are allowed to have empty values.
  - Error messages now explicitly list missing variables.
- Made the "Required" checkbox readonly in the wizard to prevent
  user modification. It is pre-defined in the server template.
- Backend improvements:
  - Added `_validate_required_variables` in the
    `cx.tower.server.template` model.
  - Validation for required variables occurs only during server creation.
  - Improved translation formatting to follow Odoo best practices
    using placeholders `%(key)s`.
- Enhanced test reliability:
  - Updated test data to avoid conflicts and ensure distinguishable references.
  - Covered edge cases for required and optional variables.

Steps to Reproduce:
1. Navigate to the "Server Templates" menu under "Server".
2. Create a new server template:
   - Fill in all required fields in the template form.
3. Under the "Configuration" tab:
   - Add several variables.
   - Mark some variables as "Required" by activating the checkbox.
4. Save the server template.

5. After saving, notice the "Create Server" button appears.
   - Click the "Create Server" button to open the wizard.

6. In the wizard:
   - Fill in all required fields for the server.
   - Observe that variables declared as "Required" in the template
     are displayed with active but non-editable checkboxes.
   - If any required variable has no value, the "Create" button
     will not be displayed.
   - If you fill in all required variable values, the "Create"
     button will become visible.

7. Attempt to create the server:
   - **Case 1:** Remove a required variable from the list of variables.
     - Upon clicking the "Create" button, an error is displayed
       from the backend:
       - "Missing variables: %(variables)s".
   - **Case 2:** Add the removed variable back but leave its value empty.
     - Attempting to create the server again results in another
       backend error:
       - "Empty values for variables: %(variables)s".

8. Once all required variables are present and filled with valid
   values, click the "Create" button.
   - The server is successfully created.

Testing:
- Automated tests for all edge cases and scenarios:
  1. test_required_flag_in_create_server_wizard
  2. test_required_attribute_in_wizard_field
  3. test_successful_server_creation_with_required_variables
  4. test_optional_variable_with_empty_value
  5. test_wizard_without_variables
  6. test_update_required_variable_value
  7. test_optional_variable_handling
- Manual testing to validate error messages and wizard behavior.

Task: 4070

ivs-cetmix committed:
[IMP] cetmix_tower_server: Store related keys

Improve the way related keys are stored in commands.
Store only keys that either global or connected to command servers
or to partners connected to command servers.

Store related secrets in the file templates.

Modify cx.tower.key constraints to ensure that they
are working as expected.
SQL constraint was useless because both 'partner' and 'server' fields
allow NULL values.
So an ORM constraint was added to ensure that the key is unique.
Delete the constrain inherited from mixin after committing database schema.

Task: 4166

nikola-vlavceski-solisa committed:
[ADD] cetmix_tower_server: add variable options selection

Add functionality to allow users to select from predefined options for variables, eliminating the need for manual value entry. This simplifies data entry and ensures consistency, especially for variables with a limited set of possible values.

Task: 4164

ivs-cetmix committed:
Update repo template to the latest version

- Apply recent changes from the OCA
- Remove tests with Odoo, use OCB only because Odoo 14.0
is not supported anymore

Commit:
Merge PR #161 into 14.0-dev

Signed-off-by ivs-cetmix

tendil committed:
[IMP] cetmix_tower_server: Centralize and enhance variable management logic

Centralized and streamlined the logic for managing `variable_ids` in `CxTowerTemplateMixin` and related models (`CxTowerCommand`, `CxTowerFile`, and `CxTowerFileTemplate`):

- Introduced `_get_relation_params` in `CxTowerTemplateMixin` to dynamically extract `relation`, `column1`, and `column2` for `variable_ids` based on field metadata.
- Created `_get_depends_fields` as a customizable method to define dependent fields for each model:
  - `CxTowerCommand` uses ["code", "path"].
  - `CxTowerFile` uses ["code", "server_dir", "name"].
  - `CxTowerFileTemplate` uses ["code", "server_dir", "file_name"].
- Enhanced `_compute_variable_ids` in `CxTowerTemplateMixin` to dynamically compute `variable_ids` using the dependencies provided by `_get_depends_fields`.
- Updated model definitions (`CxTowerCommand`, `CxTowerFile`, and `CxTowerFileTemplate`) to:
  - Retain unique relation details (`relation`, `column1`, and `column2`) for `variable_ids`.
  - Delegate computation logic to the mixin for better maintainability.
- Added professional docstrings for all methods, ensuring clear documentation and adherence to professional standards.

Additional changes:
- Adjusted UI views to include a read-only field `variable_ids`:
  - Visible only to users in the `Manager` and `Root` groups.

Task ID: 4167

Commit:
Merge PR #154 into 14.0-dev

Signed-off-by ivs-cetmix

ivs-cetmix committed:
[FIX] cetmix_tower_server: String translation

cx.tower.server.template.create.wizard:
  - Remove explicits field "string" translations.
  - Fix "compute" method name.

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Update module description

Commit:
Merge PR #164 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

tendil committed:
[IMP] cetmix_tower_server: enforce required variable validation

This commit addresses an issue where servers could be created even when all required configuration variables were removed during the server creation wizard.

Key changes:
- Enhanced `_validate_required_variables` to handle cases where all required variables are missing from the configuration:
  - If all required variables are removed, the creation process is blocked with an appropriate error message listing the missing variables.
  - The error message includes details of the missing variables for better user understanding.
- Updated `_create_new_server` to delegate validation entirely to `_validate_required_variables`, ensuring consistency and avoiding duplication of logic.

Steps to reproduce:
-------------------
1. Create a server template with at least one required variable.
2. Open the "Create Server" wizard.
3. Remove all configuration variables from the wizard.
4. Click "Confirm".

Expected result:
----------------
An error is raised, and the server is not created: "Please resolve the following issues with configuration variables:
    - Missing variables: var1, var2, var3"

Current result (before this commit):
------------------------------------
The server is created even when required variables are missing.

Testing:
--------
- Automated tests validate:
  1. Server creation with missing required variables raises an error.
  2. Removing all required variables blocks server creation.
  3. Optional variables can be left empty without affecting creation.
- Manual testing ensures:
  - Correct behavior in the "Create Server" wizard.
  - User-friendly error messages with detailed information.

Task: 4206

tendil committed:
[FIX] cetmix_tower_server: Improve test data creation and remove demo data usage

This commit addresses comments by improving the creation and usage of test data in the following ways:

- Replaced demo data usage:
  - Removed references to `demo` data (e.g., `self.env.ref("cetmix_tower_server.tag_staging")`) in test cases.
  - Introduced locally created test records for tags with unique names such as `"Test Staging"` and `"Test Production"`.

- Improved reusability:
  - Created reusable `self.Tag` model in `setUp` for tag creation in tests.
  - Moved common test data (e.g., tags) to `setUp` for easier maintenance and consistency.

- Aligned with TL recommendations:
  - Ensured test data names are unique to avoid conflicts with existing data.
  - Simplified test cases by reusing common test setup.

Why?
- Ensures independence of test cases from demo data.
- Makes tests more maintainable and aligned with best practices.
- Resolves all comments raised in the PR review.

Task: 4212

Commit:
Merge PR #168 into 14.0-dev

Signed-off-by ivs-cetmix

Aldeigja committed:
[IMP] cetmix_tower_server: expand python libraries accessible in evaluation context

- HMAC: module implements the HMAC algorithm
- HASHLIB:  hashlib module  enables hashing operations to verify data integrity during communication.

Commit:
Merge PR #158 into 14.0-dev

Signed-off-by ivs-cetmix

ivs-cetmix committed:
[ADD] cetmix_tower_git: Tower Git Management

**cetmix_tower_git:**

Base module that implements git management for Cetmix Tower.

**cetmix_tower_yaml:**

- Improve YAML generation method
- Imporve test coverage
- Update module description

Commit:
Merge PR #167 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

nikola-vlavceski-solisa committed:
[IMP] cx_tower_variable_option: implement variable option reference

Inherit from cx.tower.reference.mixin and implement the logic to generate the variable option reference.

Added the `reference` field to all relevant views, consistent with other models inheriting from this mixin.

Task: 4199

Commit:
Merge PR #159 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

nikola-vlavceski-solisa committed:
[IMP] cetmix_tower_server: SSH connection check

Added the server_check_ssh_connection function to verify the SSH connection to a server. This is useful for situations where the server is running.
The function includes parameters for server reference, retry attempts, and timeout. It returns a dictionary with the exit code and error message from the SSH check.

Task: 4196

Commit:
Merge PR #160 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

ivs-cetmix committed:
[FIX] cetmix_tower_server: revert relation fields

- Restore relation field keys removed in
  https://github.com/cetmix/cetmix-tower/pull/160/files
- Hide variables and secrets in views if they are empty.
- Remove unused function.

ivs-cetmix committed:
Update repo template to the latest version

ivs-cetmix committed:
[FIX] cetmix_tower_server: warnings in tests

Fix issues that cause warnings in tests.
Thought those are not critical, but it's better to fix them.

ivs-cetmix committed:
[FIX] cetmix_tower_yaml: warnings in tests

Add model description to the wizard.
Absent description in the wizard causes warnings in tests.

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
[UPD] Update cetmix_tower_yaml.pot

Commit:
Merge PR #177 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

nikola-vlavceski-solisa committed:
[IMP] cetmix_tower_server: Improve ui/ux

Added placeholders to the "Tags" fields in Command and Flight Plan Wizards, removed the unused "Interpreter" model, and made UI adjustments such as hiding fields for "SSH key" records and ensuring correct log formatting.

Task:4191

nikola-vlavceski-solisa committed:
[IMP] cetmix-tower-server: Variable options

- Add a new `value_char` field to the `cx.tower.variable.option` for storing values.
  This mimics behaviour of the Odoo `Selection` field.
- Update Server Template Create Wizard to use variable options properly.
- Update variable value inline form for Server and Server Template.
- Imporove test coverage

Enhance UI by adding variable options to the following views:
  - Server Template form view
  - New Server from Template wizard
  - Flight Plan action form

Updated demo data for variables:
  - Add language options with `value_char` values
  like "English (US)", "Italian", and others.
  - Updat Odoo Version options to use `value_char`
  - Enrich demo exisiting demo data to cover differen cases

Task: 4235

- Update Server Template Create Wizard to use variable options properly.
- Update variable value inline form for Server and Server Template.

Commit:
Merge PR #175 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #174 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #178 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

GabbasovDinar committed:
[FIX] cetmix_tower_server: Secrets in Python code

Python code execution fails when using a secret with a multi-line value.

Task: 4024

Commit:
Merge PR #179 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

GabbasovDinar committed:
[IMP] cetmix_tower_server: Access to server fields

Limit access to the following fields of the cx.tower.server ​omdel to Manager and Root:

- OS
- IP v4 Address
- IP v6 Address
- SSH username
- SSH Port
- SSS auth mode
- SSH key
- Use sudo

- Fix user access error when uploading a file
- Update tests

Task: 4116

GabbasovDinar committed:
[FIX] cetmix_tower_server: Server creation

- Fix incorrect creation of server variables when generating a server from a template.
- Enable the creation of servers from a template with either a full copy of variables or only the specified ones.
- Allow the creation of servers with specific variable options.
- Optimize search of existing variables and creation of new ones

Task: 4253

Commit:
Merge PR #98 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
Merge PR #170 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #180 into 14.0-dev

Signed-off-by ivs-cetmix

ivs-cetmix committed:
[FIX] cetmix_tower_server: SSH connection test

- cx.tower.server: Refactor SSH connection retrieval function.
  Use better naming, allow to pass timeout.
- cx.tower.server: Refactor SSH connection test function.
- cetmix.tower: Use SSH connection test function from cx.tower.server

tendil committed:
[IMP] cetmix_tower_server: UI & functionality updates

This commit introduces multiple UI and functionality improvements in the
`cetmix_tower_server` module, optimizing user experience and data
presentation.

Key updates:
- Sorting improvements
  - Updated Flight Plan Log and Command Log to be sorted by "Started" in
    descending order (newest first).

- Command Log enhancements
  - Added a related field to display the corresponding Flight Plan Log line
    in the Command Log form.

- Flight Plan Line form updates
  - Improved the "Code" field inside Command Preview to dynamically display:
    - File template code for the "Upload File from Template" action.
    - Related flight plan lines for the "Execute Flight Plan" action.

- Kanban view improvements
  - Displayed the number of servers in the Server Template Kanban tile using
    the `server_count` computed field.

- Search and filtering enhancements
  - Added a pre-defined "Flight Plan" filter in the Command search view.

- File Templates updates
  - Implemented an "Archive" feature for File Templates.

- Field visibility adjustments
  - The "Path" and "Sudo" fields are now only visible for commands with the
    action set to "SSH" in the Flight Plan Line form.

Task: 4263

Commit:
Merge PR #184 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

GabbasovDinar committed:
[FIX] cetmix_tower_server: Tests

- Fix tests falling due to missing variable.

Task: 4273

Commit:
Merge PR #186 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

ivs-cetmix committed:
[IMP] cetmix_tower_server: On delete flight plan

Execute flight plan when server is deleted.

Set "Deleting" status when flight plan is started.
Delete server if plan is successful.
Set "Delete Error" status when flight plan fails.

GabbasovDinar committed:
[FIX] cetmix_tower_server: Server from template

- Fix using server configuration when creating a server from a template

Task: 4272

GabbasovDinar committed:
[REF] cetmix_tower_server: Tests

Use MagicMock for mocking SSH and file operations.
This is done to make the test process more clean and clear.

Task: 4116

Commit:
Merge PR #176 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
[UPD] Update cetmix_tower_git.pot

Commit:
[UPD] Update cetmix_tower_yaml.pot

Commit:
Merge PR #190 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
Merge PR #189 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #182 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #183 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] repository files

- Update CONTRIBUTING.md
- Use CLA assistant for signing ICLA.
- Add issue templates

GabbasovDinar committed:
[IMP] cetmix_tower_server: Server from Template

- Add default ondelete Flight Plan for new servers from Server Template

Task: 4293

ivs-cetmix committed:
Add PR template

Commit:
Merge PR #191 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: New models and fields

**Variables**

- Variable Type
- Options

**Keys**

- Don't export secret value, remove obsolete methods.

**Server Template**

- Export 'On Delete Plan'.

**Export Wizard**

- Option to remove empty x2m field values. This is useful for keeping
  the YAML file smaller and cleaner.
- Improve form formatting for better UX.

Task: 4219

ivs-cetmix committed:
[IMP] cetmix_tower_server Server URL

Add new 'URL' field to improve server operations UX.
Add new 'tower.server.url' system variable.
Refactor Server kanban view: add elements based on condition instead of
hiding them.

Task: 4301

Commit:
Merge PR #195 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #193 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_yaml.pot

ivs-cetmix committed:
Update README.md

Signed-off-by: Ivan Sokolov <2234999+ivs-cetmix@users.noreply.github.com>

ivs-cetmix committed:
[IMP] cetmix_tower_server: Variable values

Replace dynamic domain with a static one for variable values.
This reduces code complexity and improves performance.

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Variable values

Export related variables of variable values.

ivs-cetmix committed:
[IMP] cetmix_tower_server: Python command

Add 'delete' method to requests library used in Python commands.

Commit:
Merge PR #196 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[FIX] cetmix_tower_server_notify_backend On delete

When a Server is deleted an on delete Flight Plan is triggered.
As soon as this Flight Plan finishes with success,
the Server is deleted too. All related records are deleted with it.

This creates an issue when notification is sent for a log record
that is already deleted.

This commit fixes this issue by checking if the record exists
before sending a notification.

Task: 4320

mymage committed:
[i18n] updated it.po for all modules

Commit:
Merge PR #197 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

ivs-cetmix committed:
[FIX] cetmix_tower_server: Variable value, Server Template & Logs

**Variable Values**

Currently "Manager" group members can only create variable values
for servers they are followers of.
However, partners are added as followers of servers when they are
created.
This commit adds access to variable values created by users.

NB: this is a temporary fix. To be replaced by a more generic
access control system.

**Server Template & Logs**

Server Template: use `sudo` when preparing server values to ensure
that all values are copied.

Server Logs: add security rules to allow users to access their own logs and
logs of servers/server templates they have access to.

Task: 4323

ivs-cetmix committed:
[FIX] cetmix_tower_yaml: YAML Import

Remove constraint on `yaml_code` field because it raises exception
even when user is not accessing the field explicitly.
This blocked non members of `cetmix_tower_yaml.group_import` group
from creating and modifying records in models inheriting from
`cx.tower.yaml.mixin`.

ivs-cetmix committed:
[IMP] cetmix_tower_server: New access rules

Before this commit access rules we based on followers.
This is more Odoo style, however this creates issues, because
followers are added after record is created. So such rules cannot
be used to regulate "create" access.
Also this doesn't allow to define granular access for users, where
some users have other access rules while belonging to the same group.
Eg one would like to give read access while maintaining record modification
restriction.

This commit implemens a new mixin that adds two fields to models inheriting from it:
"Users" and "Managers".
It allows to define access rules based on those fields, providing more flexible way
to restrict access to records.
Eg now it's possible to give some users read-only access to Server or Server Template
records while letting others to modify them.

Please check the module documentation for details.

**Breaking changes**:

Must add existing users of the "Manager" as Managers of existing Servers
in case it's needed to keep write access to those records.

Task: 4327

Commit:
Merge PR #198 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
Merge PR #199 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

ivs-cetmix committed:
[IMP] cetmix_tower_git: New access rules

New access rules according to the new access rules for the server.

Commit:
Merge PR #200 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
[UPD] Update cetmix_tower_git.pot

ivs-cetmix committed:
[FIX] cetmix_tower_server: Related records creation

When a record is created, some related records are created as well.
Some of them use rules that depend on `user_ids` and `manager_ids` fields.
However at the moment of record creation, these fields are not yet set.
So first we create the record without these related records, then we create
the related records to avoid access violations.

Task: 4369

ivs-cetmix committed:
[FIX] cetmix_tower_git: Related records creation

When a record is created, some related records are created as well.
Some of them use rules that depend on `user_ids` and `manager_ids` fields.
However at the moment of record creation, these fields are not yet set.
So first we create the record without these related records, then we create
the related records to avoid access violations.

Task: 4369

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: add new fields

Variables: access level
Variable Values: access level

Commit:
Merge PR #185 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] cetmix_tower_server: Access rules for keys

Keys/Secrets: access rules based on access roles.

Task: 4300

Commit:
Merge PR #201 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #202 into 14.0-dev

Signed-off-by ivs-cetmix

ivs-cetmix committed:
[IMP] cetmix_tower_server: UI/UX updates

Server Form: re-arrange "Reference" and "URL"fields.
Server Template Form: re-arrange "Reference" and "URL"fields.
Command Execute Wizard: new design for better UX.
Plan Execute Wizard: new design for better UX.
Server: check if command can be run on server before running it.
Flight Plan: check if flight plan can be run on server before running it.

Commit:
Merge PR #204 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #203 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

GabbasovDinar committed:
[IMP] cetmix_tower_server: Refactor SSH handling

- Move SSH library to a separate folder.
- Refactore code for improved scalability and maintainability using SOLID principles in the SSH class.
- Implemente connection caching to avoid repeatedly opening new connections.
- Add a decorator to automatically close connections after transaction commit or rollback, preventing accumulation of open connections.
- Rename type of ssh_port field from Char to Integer

Task: 4142

Commit:
Added translation using Weblate (German)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (483 of 483 strings)

Translation: cetmix-tower-14.0-dev/cetmix_tower_server
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_server/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (80 of 80 strings)

Translation: cetmix-tower-14.0-dev/cetmix_tower_yaml
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_yaml/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (85 of 85 strings)

Translation: cetmix-tower-14.0-dev/cetmix_tower_git
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_git/it/

AquaMCU committed:
Translated using Weblate (German)

Currently translated at 4.1% (21 of 508 strings)

Translation: cetmix-tower-14.0-dev/cetmix_tower_server
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_server/de/

ivs-cetmix committed:
[FIX] cetmix_tower_server: Create server from template

Error is raised if user doesn't have access to a file template used
is Server Template logs.

Commit:
Merge PR #208 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
Merge PR #207 into 14.0-dev

Signed-off-by ivs-cetmix

ivs-cetmix committed:
[IMP] cetmix_tower_server: New system variables

Currently {{ tower.tools.now }} with return date as date time
string, eg '2025-02-28 21:58:31'.
Such string is not so usable inline because it contains various symbols
including spaces.
Same story with the {{ tower.tools.date }} variable.

Add new system variables:

- {{ tower.tools.now_underscore }}
- {{ tower.tools.date_underscores }}

They return date and datetime values with underscores as separators.
Eg '2025_02_28_21_58_31'.

ivs-cetmix committed:
[IMP] Export related secrets

cetmix_tower_server
-------------------

Improve the way 'today_underscore' and 'now_underscore' are computed.

cetmix_tower_yaml
-----------------

Export related  are computed.

cetmix_tower_yaml
-----------------

- Export related secrets.
- Show secrets used in code in yaml import wizard.
  Helps user to see which secrets may need his attention
  after import is completed.

ivs-cetmix committed:
Update template repo to the latest version

ivs-cetmix committed:
[FIX] cetmix_tower_server: File access rules

Manager cannot read files if added only to "Managers" in
the related server.

Commit:
Merge PR #211 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

badbole committed:
Added translation using Weblate (Croatian)

Translated using Weblate (Croatian)

Currently translated at 90.5% (460 of 508 strings)

Translated using Weblate (Croatian)

Currently translated at 39.7% (202 of 508 strings)

Translated using Weblate (Croatian)

Currently translated at 9.4% (9 of 95 strings)

Added translation using Weblate (Croatian)

Added translation using Weblate (Croatian)

Co-authored-by: Bole <bole@dajmi5.com>
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_git/hr/
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_server/hr/
Translation: cetmix-tower-14.0-dev/cetmix_tower_git
Translation: cetmix-tower-14.0-dev/cetmix_tower_server

Commit:
Translated using Weblate (German)

Currently translated at 4.1% (21 of 508 strings)

Co-authored-by: CetmixWeblateBot <weblate@cetmix.com>
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_server/de/
Translation: cetmix-tower-14.0-dev/cetmix_tower_server

jusle committed:
Added translation using Weblate (Finnish)

Added translation using Weblate (Finnish)

Co-authored-by: Jussi Lehto <jussi.lehto@cetmix.com>

ivs-cetmix committed:
[IMP] cetmix_tower_server: UI/UX updates

 - Flight plan log: Sort command logs in flight plan log from latest to
   newest. Currently it's from newest to oldest.
 - Variable: Show "option" in the "Values" tab.
 - Fix form view for variables (options).
 - Variable value: sequences for ability to arrange the lines manually
 - Flight Plan action: use "Action" field for command domain, improve
   the plan action view.
 - Restrict command deletion if used in flight plan lines.

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: UI/UX updates

- Variable value: export sequences.

Commit:
Merge PR #210 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
Merge PR #209 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_yaml.pot

Commit:
Merge PR #213 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Import & export wizards

Import wizard:

- Always use existing secrets/keys when importing YAML
- Reformat import wizard form for better user experience

Export wizard:

- Reformat export wizard form for better user experience

Task: 4399

ivs-cetmix committed:
[IMP] cetmix_tower_server: Demo data

Update names and references for demo data to avoid potential conflicts
with existing records.

Task: 4399

ivs-cetmix committed:
[IMP] cetmix_tower_git: UI/UX improvements

- Git Project File Link: show "open" button in Server o2m field
- Git Project File Link: show "Auto Sync" (related to file)
- Git Project: unlink file from file template when linking to a project.
- Git Project: show create_uid (optional=hide) for sources and remotes
- Link Git Project to File Template
- Git Remote: Set default project format
- Git Remote: Use regex to detect git ssh urls
- Server: add 'Open' button to project line

Task: 4402

Commit:
Merge PR #216 into 14.0-dev

Signed-off-by ivs-cetmix

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 92.5% (74 of 80 strings)

Translated using Weblate (Croatian)

Currently translated at 93.5% (475 of 508 strings)

Translated using Weblate (Croatian)

Currently translated at 47.3% (45 of 95 strings)

Translated using Weblate (Croatian)

Currently translated at 81.2% (65 of 80 strings)

Translated using Weblate (Croatian)

Currently translated at 93.3% (474 of 508 strings)

Translated using Weblate (Croatian)

Currently translated at 26.3% (25 of 95 strings)

Translated using Weblate (Croatian)

Currently translated at 93.3% (474 of 508 strings)

Translated using Weblate (Croatian)

Currently translated at 20.0% (16 of 80 strings)

Translated using Weblate (Croatian)

Currently translated at 90.7% (461 of 508 strings)

Co-authored-by: Bole <bole@dajmi5.com>
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_git/hr/
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_server/hr/
Translate-URL: https://hosted.weblate.org/projects/tower-server-14-0-dev/cetmix_tower_yaml/hr/
Translation: cetmix-tower-14.0-dev/cetmix_tower_git
Translation: cetmix-tower-14.0-dev/cetmix_tower_server
Translation: cetmix-tower-14.0-dev/cetmix_tower_yaml

Commit:
[UPD] Update cetmix_tower_yaml.pot

Commit:
Merge PR #206 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #215 into 14.0-dev

Signed-off-by ivs-cetmix

halbtonjazz committed:
[IMP] cetmix_tower_server: unify sudo handling

Replace sudo selector with a checkbox in command execution interfaces to
ensure consistent behavior with server settings. This prevents potential
deadlocks when server sudo mode (with password) conflicts with command
execution mode (without password).

Key changes:
- Replace Selection field with Boolean for sudo mode configuration
- Unify sudo logic across command wizard and flight plans

This change simplifies sudo handling and prevents situations where
tower could hang waiting for password input while executing commands.

Task: 4403

ivs-cetmix committed:
[IMP][BREAKING] cx_tower_server: Refactor secrets

Before this commit
-------------------

Current implementation of secrets presumes that a secret can be linked
either to a server or a partner.
This requires to make an exclusion from the unique constraint for the
references to allow to have several secrets with the same reference.
This breaks the general paradigm of references being unique, raises
confusion and is error prone.

After this commit
-----------------

New dedicated model is added to store custom secret values for secrets.
At the same time backward compatibility is kept by converting "Secret Value"
field of a secret into a computed one.

Now there is no need in custom reference generation for secrets.
Also it's much simpler to understand and maintain secrets logic and
compute "Related Secrets" field in Commands, Files and File Templates.

Task: 4413

Commit:
Merge PR #219 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_git.pot

GabbasovDinar committed:
[IMP] cetmix_tower_key: SSH Server Auth Policy

Implement host key auth policies to improve connection security.
User can not provide a host key, or opt for not checking it explicitly.
Host key can be also obtained automatically via password protected wizard.

Task: 4398

- Fix pyling warnings

ivs-cetmix committed:
[IMP] cetmix_tower_server: New server wizard

Add the following fields to the new server wizard to improve the UX:

- Partner
- Operating System
- Tags
- Use sudo

Task: 4262

ivs-cetmix committed:
[IMP] cetmix_tower_server: Variable form buttons

Show related records in variables so users can navigate to them easily
directly from the variable form.

Task: 4445

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: multi record support

Before
------

Records were exported and imported one by one.
This was not so handy, especially when importing a lot of records or
handling complex configurations.

After
-----

Multiple records can be exported and imported at once.
Moreover, records from different models can be mixed in the same file
and imported all together.
This adds flexibility and improves the UX.
Add new "skip" option in import wizard to allow skipping existing records.

Task: 4446

Commit:
Merge PR #228 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #227 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #226 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #224 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #225 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] cetmix_tower_server: Host key in templates

FIX
---

Update migration scripts from previous PRs.

UI/UX
-----

- Server from Template wizard: add "host key" and "skip host key" fields
- Server Template: re-arrange ip address, "Tags" and "Color" fields
- Server: re-arrange "Tags" and "Color" fields

Task: 4457

Commit:
Merge PR #230 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[FIX] cetmix_tower_server: Migration script

Fix error in SQL in the migration script that saves global key values.

Commit:
Merge PR #231 into 14.0-dev

Signed-off-by ivs-cetmix

GabbasovDinar committed:
[IMP] cetmix_tower_server: Post migration

- Set default skip host key for exists servers
- Fix error retrieving host key after click "Get from Host" button

Task: 4461

ivs-cetmix committed:
[IMP] cetmix_tower_server: UI/UX improvements

Server
--------

Allow to select "secret" type key only in the Server "Secrets" tab.

Server Log
-----------

Server log: allow to select only ssh and python commands for 'command' type logs

Command
----------

- Help is computed field now. This simplifies inheritance and extension
  when added new features.
- Show flight plans command is used in
- Rename `execute_command` into `run_command`:

**Why?**

Top level methods are often used in other modules and Odoo automation.
So we should keep their naming aligned with general naming recommendations.
Because we are using `Run Command` in UI, we should rename the method into
`run_command` too to keep UI and method names aligned.

Flight Plan
-----------

- Add `run_flight_plan` method to `cx.tower.server` model.
- Remove `execute` method from `cx.tower.plan` model:

**Why?**

We already have `run_command` method in `cx.tower.server` model.
However to run a flight plan before this updated user had to use `execute`
method from `cx.tower.plan` model.
Now user can use `run_flight_plan` method from `cx.tower.server` model,
so it makes code more consistent and easier to understand.
At the same time `execute` method from `cx.tower.plan` model is not needed
anymore and can be removed.

Files
-----

- Rename 'action_modify_code'. New name: 'action_unlink_from_template'
  to make it more clear
- Set default cron job for pulling files from server to 1 day
- Add general settings for Tower Server
- Update menu items
- Allow using '/' as dir in file path to work with files located in root dir

OS
---

- Update access rights and views.
- Add OS compatibility warning to command execute wizard.
- Update and improve the demo data.

Tag
---

- Add related fields on the tag side
- Add security rules and checks for tags

Task: 4466

Commit:
Merge PR #232 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
Merge PR #229 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_yaml.pot

ivs-cetmix committed:
[IMP] cetmix_tower_server: variable value modifier

Allow to modify variable values using Python code.

Why?
----

In case we need to post-process the variable value, for example to
remove special characters or modify it based on some logic.
Eg. we have a variable which values are used for URLs, so we want to
remove all special characters from the value.

Task: 4474

Commit:
Merge PR #234 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[FIX] cetmix_tower_server: Variable modifier

Fix issue when variable modifiers were not applied to variables used
in another variable values.
Eg if for a variable `sentence` with value `{{ word_1 }} + {{ word_2}}`
modifiers were applied only to `sentence` but not to `word_1` and
`word_2`.

Task: 4490

GabbasovDinar committed:
[FIX] cetmix_tower_server: Required host key

- The host key is required in form view

Task: 4473

Commit:
Merge PR #236 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #233 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
[UPD] Update cetmix_tower_yaml.pot

ivs-cetmix committed:
[IMP] cetmix_tower_server: Shortcuts

Shortcuts allow to run commands or flight plans with a single click.
Shortcuts can be assigned to servers or server templates.

Task: 4491

ivs-cetmix committed:
[IMP] cetmix_tower_server: Variable validation

Add regex validation for variable.
This is done to ensure that only correct values are saved in the database.

Task: 4505

Commit:
Merge PR #237 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
Merge PR #238 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[UPD] Update cetmix_tower_server.pot

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: New models for export

- Export the "Shortcuts" model
- Export new fields of the "Variable" model:
  - Applied Expression
  - Validation Pattern
  - Validation Message
- Export new fields of the "Server Template" model:
  - Shortcuts

Task: 4506

Commit:
Merge PR #239 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_yaml.pot

Commit:
Merge PR #235 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

ivs-cetmix committed:
[IMP] cetmix_tower_server: UI/UX updates

- Files: refactor view to make it more user-friendly
- Shortcuts: add 'Note' field, refactor views, fix tests

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Add the 'note' field of the Shortcuts to export.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_git: UI/UX improvements

- Add 'Note' field to git project
- Clean up code before merging into 14.0

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_notify_backend Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server_queue Code cleanup

Cleanup code to comply with the OCA coding guidlines.
Fix minor issues.

Task: 4509

ivs-cetmix committed:
[FIX] Variable Values

- Fix variable value check
- Reset option_id when value_char is invalid
- Clear value_char when option_id is removed

ivs-cetmix committed:
[FIX] Server Host Key

Do not check host key if IP address is not present.
This may happen when server is created from template and IP address is
assigned by a flight plan that runs after server creation.

ivs-cetmix committed:
[ADD] cetmix_tower App Store Module

This module is a collection of tools for the Tower project.
It's main purpose is to be used in the Odoo App Store so users could
get all modules at once without searching for them.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server: Command result

Replace 'COMMAND_RESULT' with 'result' in command 'code' field.
This is done to align with the general Odoo practice of using 'result'
and to keep the same coding style across the codebase.
Migration script is added to update existing commands.

Task: 4509

ivs-cetmix committed:
[IMP] cetmix_tower_server: command timeout

Check 'zombie' commands and mark them as finished.
This is needed to avoid infinite loops when a command is stuck.

Task: 4509

ivs-cetmix committed:
Fix directory structure

Commit:
Merge PR #240 into 14.0-dev

Signed-off-by ivs-cetmix

Commit:
[BOT] post-merge updates

Commit:
[UPD] Update cetmix_tower_git.pot

Commit:
[UPD] Update cetmix_tower.pot

Commit:
[UPD] Update cetmix_tower_server.pot

Commit:
[UPD] Update cetmix_tower_yaml.pot

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cetmix/cetmix-tower?pullRequest=241) <br/>All committers have signed the CLA.

ivs-cetmix committed:
Merge branch '14.0' into 14.0-dev

Signed-off-by: Ivan Sokolov <2234999+ivs-cetmix@users.noreply.github.com>

ivs-cetmix commented:
@coderabbitai full review

ivs-cetmix committed:
[IMP] Update modules maturity level

Set all modules to the `Stable/Production` maturity level.
Bump `cetmix_tower_server` version to `14.0.1.0.0`.

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix commented:
Let's go!
/ocabot merge nobump

CetmixGitDrone commented:
This PR looks fantastic, let's merge it!
Prepared branch [14.0-ocabot-merge-pr-241-by-ivs-cetmix-bump-nobump](https://github.com/cetmix/cetmix-tower/commits/14.0-ocabot-merge-pr-241-by-ivs-cetmix-bump-nobump), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 70073bba0a1b24a362690842c22a4340bc5d4ca3. Thanks a lot for contributing to cetmix. ❤️
